tools/isobuild/import-scanner.js

```diff
@@ -0,0 +1,431 @@
+
+    if (installPath.startsWith("..")) {
+      // absPath is not a subdirectory of this.sourceRoot.
+      return;
+    }
+
+    const dirs = this._splitPath(pathDirname(installPath));
+    const bundlingClientApp =
+      ! this.name && // Indicates we are bundling an app.
+      archMatches(this.bundleArch, "web");
+
+    for (let dir of dirs) {
+      if (dir.charAt(0) === "." ||
+          dir === "packages" ||
+          dir === "programs" ||
+          dir === "cordova-build-override") {
+        // These directories are never loaded as part of an app.
+        return;
+      }
+
+      if (bundlingClientApp && (dir === "server" ||
+                                dir === "private")) {
+        // If we're bundling an app for a client architecture, any files
+        // contained by a server-only directory that is not contained by
+        // a node_modules directory must be ignored.
+        return;
+      }
+
+      if (dir === "node_modules") {
+        if (! this._hasKnownExtension(installPath)) {
+          // Reject any files within node_modules directories that do
+          // not have one of the known extensions.
+          return;
+        }
+
+        // Accept any file within a node_modules directory if it has a
+        // known file extension.
+        return {
+          path: installPath,
+          inNodeModules: true,
+        };
+      }
+    }
+
+    return {
+      path: installPath,
+      inNodeModules: false,
+    };
+  }
+
+  _hasKnownExtension(path) {
+    return this.extensions.indexOf(pathExtname(path)) >= 0;
+  }
+
+  _splitPath(path) {
+    const partsInReverse = [];
+    for (let dir; (dir = pathDirname(path)) !== path; path = dir) {
+      partsInReverse.push(pathBasename(path));
+    }
+    return partsInReverse.reverse();
+  }
+
+  _tryToResolveImportedPath(id, path) {
+    let resolved =
+      this._resolveAbsolute(id) ||
+      this._resolveRelative(id, path) ||
+      this._resolveNodeModule(id, path);
+
+    while (resolved && resolved.stat.isDirectory()) {
+      resolved = this._resolvePkgJsonMain(resolved.path) ||
+        this._joinAndStat(dirPath, "index.js");
+    }
+
+    return resolved && resolved.path;
+  }
+
+  _joinAndStat(...joinArgs) {
+    const path = pathNormalize(pathJoin(...joinArgs));
+    const exactStat = statOrNull(path);
+    const exactResult = exactStat && { path, stat: exactStat };
+    if (exactResult && exactStat.isFile()) {
+      return exactResult;
+    }
+
+    for (let ext of this.extensions) {
+      const pathWithExt = path + ext;
+      const stat = statOrNull(pathWithExt);
+      if (stat) {
+        return { path: pathWithExt, stat };
+      }
+    }
+
+    if (exactResult && exactStat.isDirectory()) {
+      // After trying all available file extensions, fall back to the
+      // original result if it was a directory.
+      return exactResult;
+    }
+
+    return null;
+  }
+
+  _resolveAbsolute(id) {
```

Comment:
An important concept in this implementation is that `meteorInstall` does not have to preserve the exact structure of files and directories from the developer's file system. Instead, `meteorInstall` has the opportunity to define a simplified directory structure that allows the usual Node lookup algorithm to work as expected, without a bunch of special cases. You can think of `_getInstallInfo` as a method that maps real file system paths to normalized virtual `.installPath`s for use with `meteorInstall`. Sometimes `_getInstallInfo` will reject a path, which just means that module won't be installed by `meteorInstall`, and can't be imported at runtime (for example, any file in a `server/` directory when we're building a bundle for the client).

Given the current implementation, the virtual directory structure expressed via `meteorInstall` might look something like this:
```
<root>
├── app # same for all apps, allows apps to have their own local node_modules directory
│   ├── eager1.js # a file that will be required immediately when the app starts
│   ├── eager2.js
│   ├── imports
│   │   ├── lazy1.js # a file that must be imported
│   │   └── lazy2.js
│   └── node_modules
│       └── moment # a local npm package installed by the app developer
└── node_modules
    ├── pkg1 # a Meteor package with two npm dependencies
    │   └── node_modules
    │       ├── npm1 # in reality, the source files for these npm packages live in some .npm or 
    │       └── npm2 # isopacket directory, but _getNodeModulesInstallInfo normalizes them thus
    ├── pkg2 # a Meteor package with one npm dependency
    │   └── node_modules
    │       └── npm3
    └── pkg3.js # generated stub for a Meteor package not using modules: module.exports = Package.pkg3
```

What to I mean by special cases? For example, the actual source files for that `npm3` dependency might reside in different places depending on whether the `pkg2` Meteor package was installed from a checkout, was in a local `packages/` directory, or came from Atmosphere, so it's vitally important that `meteorInstall` always installs `pkg2` as `/node_modules/pkg2/` and the `npm3` dependency as `/node_modules/pkg2/node_modules/npm3/`.

Given this structure, it is technically possible to `import "/app/imports/lazy2"`, though I would be willing to consider disabling support for absolute (`/`) identifiers, which would effectively force all imported identifiers to be either relative (`./` or `../`) or top-level (i.e., referring to something in `node_modules`).

Even if we forbid absolute identifiers, it would still be possible for packages to use `../../app/` relative identifiers to import things from app code. If we wanted to forbid that, we could depart from the Node lookup algorithm by inventing a new special case, or give the `/app` directory a randomized name, but I don't think either of those options is really worth the added complexity. Node developers can explore the entire file system using relative identifiers, but no one is so foolish as to do that in real code.